### ISSUE TITLE
AccessPolicy recreates a remote policy after losing its status checkpoint

### ISSUE THREAD

**Issue by isac322** [isac322/flareway#56]
## Summary

When `AccessPolicyReconciler` manages a new policy it calls `CreateAccessPolicy` first and only afterwards records the returned remote ID in `status.policyId` via `patchAccessPolicyStatus`. There is no pre-create lookup by the ownership-marker name (`flareway/<cluster>/<ns>/<name>`) and no durable checkpoint written before the remote create. If the status write never lands — an apiserver error, a conflict, or a controller restart/eviction between the create and the patch — the next reconcile sees an empty `status.policyId` and calls `CreateAccessPolicy` again. Cloudflare permits duplicate policy names, so the first remote policy is orphaned: it is never tracked, never updated, and never deleted because `reconcileDelete` keys solely on `status.policyId`.

The sibling `AccessGroupReconciler` already survives this failure mode: it persists the identity with `persistIdentity` under `retry.RetryOnConflict` and recovers an ambiguous create through `findAccessGroupByName`.

## Impact

Every lost checkpoint leaks one orphaned Cloudflare Access policy per AccessPolicy. Orphaned policies may remain attached to Access applications in Cloudflare with no Kubernetes owner, accumulate on each repeated loss, and are invisible to `kubectl` and to the controller's own deletion path.

## Tested revision and evidence limits

Verified on commit `ca23ba86f738214cc9ea7393656a74ef0782ace9` (current `main`) with a unit-level probe: the real reconciler and controller-runtime fake client plus a fake Cloudflare API boundary. No live Cloudflare or Envoy end-to-end environment was used; the evidence covers the in-cluster control-plane path only. The probe injects the checkpoint loss directly (clearing `status.policyId` after a successful first reconcile) rather than simulating a specific patch-conflict mechanism, so the confirmed scope is: any failure to persist `status.policyId` after `CreateAccessPolicy` produces a duplicate remote create.

## Reproduction

The attached test file adds a diagnostic probe to `internal/controller`. It reconciles a managed AccessPolicy once, clears `status.policyId` to model the lost checkpoint, reconciles again, and counts remote creates. The probe **passes while the defect reproduces** (two remote policies with the same ownership-marker name) and fails once the controller recovers the orphan instead of creating a second policy.

Fresh disposable checkout of `isac322/flareway` at the tested revision; no Cloudflare credentials or live cluster required for this component-level probe.

```sh
git clone https://github.com/isac322/flareway.git flareway-repro
cd flareway-repro
git checkout --detach ca23ba86f738214cc9ea7393656a74ef0782ace9
```

Add the following file(s) to that checkout (existing repository test fixtures are reused where noted in the file).

**`internal/controller/issue_accesspolicy_checkpoint_test.go`**

```go
// Diagnostic probe: it PASSES while the defect reproduces and fails once the
// controller recovers the orphaned remote policy instead of creating a second
// one.
//
// AccessPolicyReconciler creates the remote policy before any durable identity
// checkpoint and never searches for an existing remote policy by its
// ownership-marker name. When the persisted status.policyId is lost -- an
// apiserver error, a conflict, or a controller restart between
// CreateAccessPolicy and the status patch -- the next reconcile creates a
// second remote policy and orphans the first.
package controller

import (
	"context"
	"fmt"
	"testing"

	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/types"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"

	v1alpha1 "github.com/isac322/flareway/api/v1alpha1"
	flarecloudflare "github.com/isac322/flareway/internal/cloudflare"
)

func issuePolicyScheme(t *testing.T) *runtime.Scheme {
	t.Helper()
	scheme := runtime.NewScheme()
	if err := corev1.AddToScheme(scheme); err != nil {
		t.Fatal(err)
	}
	if err := v1alpha1.AddToScheme(scheme); err != nil {
		t.Fatal(err)
	}
	return scheme
}

// issuePolicyFixture seeds an authorized tenant namespace, the cluster-id
// namespace, the credential Secret, and an accepted CloudflareAccount.
func issuePolicyFixture(t *testing.T, objects ...client.Object) (client.Client, *runtime.Scheme) {
	t.Helper()
	scheme := issuePolicyScheme(t)
	labels := map[string]string{"issue-tenant": "true"}
	seed := []client.Object{
		&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "tenant-a", Labels: labels}},
		&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "kube-system", UID: types.UID("issue-cluster")}},
		&corev1.Secret{
			ObjectMeta: metav1.ObjectMeta{Name: "api-token", Namespace: "tenant-a"},
			Data:       map[string][]byte{"token": []byte("value")},
		},
		&v1alpha1.CloudflareAccount{
			ObjectMeta: metav1.ObjectMeta{Name: "account"},
			Spec: v1alpha1.CloudflareAccountSpec{
				AccountID: "0123456789abcdef0123456789abcdef",
				Credentials: v1alpha1.CloudflareAccountCredentials{APITokenSecretRef: v1alpha1.NamespacedSecretKeyReference{
					Name: "api-token", Namespace: "tenant-a", Key: "token",
				}},
				Grants: []v1alpha1.CloudflareAccountGrant{{
					NamespaceSelector: metav1.LabelSelector{MatchLabels: labels},
					Hostnames:         []string{"*"},
					Zones:             []string{"*"},
					Exposures:         []v1alpha1.Exposure{v1alpha1.ExposurePublic, v1alpha1.ExposurePrivate},
					AccessPolicyRefs:  v1alpha1.GrantPermissionAllowed,
					PlatformObjects:   v1alpha1.GrantPermissionAllowed,
				}},
			},
			Status: v1alpha1.CloudflareAccountStatus{Conditions: []metav1.Condition{
				{Type: v1alpha1.CloudflareAccountConditionAccepted, Status: metav1.ConditionTrue, Reason: "Accepted", LastTransitionTime: metav1.Now()},
				{Type: v1alpha1.CloudflareAccountConditionCredentialsValid, Status: metav1.ConditionTrue, Reason: "Valid", LastTransitionTime: metav1.Now()},
			}},
		},
	}
	seed = append(seed, objects...)
	kube := fakeclient.NewClientBuilder().WithScheme(scheme).
		WithStatusSubresource(&v1alpha1.AccessPolicy{}).
		WithObjects(seed...).Build()
	return kube, scheme
}

// issuePolicyAPI is a stateful fake of the Access policy surface the
// reconciler uses. Cloudflare permits duplicate policy names, which is what
// makes the orphan observable.
type issuePolicyAPI struct {
	flarecloudflare.AccessAPI
	policies map[string]flarecloudflare.AccessPolicy
	creates  int
	next     int
}

func (f *issuePolicyAPI) CreateAccessPolicy(_ context.Context, input flarecloudflare.AccessPolicyInput) (flarecloudflare.AccessPolicy, error) {
	f.next++
	f.creates++
	policy := flarecloudflare.AccessPolicy{
		ID: fmt.Sprintf("remote-policy-%d", f.next), Name: input.Name,
		Decision: input.Decision, Include: input.Include,
	}
	f.policies[policy.ID] = policy
	return policy, nil
}

func (f *issuePolicyAPI) GetAccessPolicy(_ context.Context, id string) (flarecloudflare.AccessPolicy, error) {
	policy, found := f.policies[id]
	if !found {
		return flarecloudflare.AccessPolicy{}, fmt.Errorf("fake: policy %q not found", id)
	}
	return policy, nil
}

func (f *issuePolicyAPI) ListAccessPolicies(context.Context) ([]flarecloudflare.AccessPolicy, error) {
	out := make([]flarecloudflare.AccessPolicy, 0, len(f.policies))
	for _, policy := range f.policies {
		out = append(out, policy)
	}
	return out, nil
}

func TestIssueAccessPolicyRecreatesRemotePolicyAfterCheckpointLoss(t *testing.T) {
	ctx := context.Background()
	policy := &v1alpha1.AccessPolicy{
		ObjectMeta: metav1.ObjectMeta{
			Name: "checkpoint-loss", Namespace: "tenant-a",
			Finalizers: []string{v1alpha1.AccessPolicyFinalizer},
		},
		Spec: v1alpha1.AccessPolicySpec{
			AccountRef: corev1.LocalObjectReference{Name: "account"},
			Name:       "checkpoint-loss",
			Decision:   v1alpha1.AccessPolicyDecisionAllow,
			Include:    []v1alpha1.AccessRule{{Everyone: &v1alpha1.AccessEveryoneRule{}}},
		},
	}
	kube, scheme := issuePolicyFixture(t, policy)
	api := &issuePolicyAPI{policies: map[string]flarecloudflare.AccessPolicy{}}
	reconciler := &AccessPolicyReconciler{
		Client: kube, Scheme: scheme,
		NewCloudflareClient: func(string, string) (flarecloudflare.AccessAPI, error) { return api, nil },
	}
	key := client.ObjectKeyFromObject(policy)
	if _, err := reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: key}); err != nil {
		t.Fatalf("first reconcile: %v", err)
	}
	if api.creates != 1 {
		t.Fatalf("CreateAccessPolicy calls after the first reconcile = %d, want 1", api.creates)
	}
	var current v1alpha1.AccessPolicy
	if err := kube.Get(ctx, key, &current); err != nil {
		t.Fatal(err)
	}
	firstID := current.Status.PolicyID
	if firstID == "" {
		t.Fatal("the first reconcile did not record status.policyId")
	}

	// Inject the checkpoint loss: the status write that recorded policyId
	// never landed (apiserver error, conflict, or controller restart between
	// CreateAccessPolicy and the status patch).
	current.Status.PolicyID = ""
	current.Status.OwnershipVerified = false
	if err := kube.Status().Update(ctx, &current); err != nil {
		t.Fatal(err)
	}

	if _, err := reconciler.Reconcile(ctx, ctrl.Request{NamespacedName: key}); err != nil {
		t.Fatalf("second reconcile: %v", err)
	}
	if api.creates != 2 {
		t.Fatalf("defect absent: CreateAccessPolicy calls after checkpoint loss = %d, want 2", api.creates)
	}
	name := api.policies[firstID].Name
	matching := 0
	for _, remote := range api.policies {
		if remote.Name == name {
			matching++
		}
	}
	if matching != 2 {
		t.Fatalf("remote policies named %q = %d, want 2 (one orphaned)", name, matching)
	}
	t.Logf("defect reproduced: remote policy %q is orphaned; a lookup by the ownership-marker name %q could have recovered it", firstID, name)
}
```

Run:

```sh
go test -mod=readonly -count=1 -v -run 'TestIssueAccessPolicyRecreatesRemotePolicyAfterCheckpointLoss' github.com/isac322/flareway/internal/controller
```

Executed against the commit above with Go 1.27.1: exit code `0`.
This is a diagnostic probe: it passes because it observes the defective behavior described here (plus the stated control cases). It is not a passing regression test.

Assertion output from that run:

```text
issue_accesspolicy_checkpoint_test.go:182: defect reproduced: remote policy "remote-policy-1" is orphaned; a lookup by the ownership-marker name "flareway/issue-cluster/tenant-a/checkpoint-loss" could have recovered it
```

## Actual result

`CreateAccessPolicy` is invoked twice for the same AccessPolicy and the fake remote holds two policies with the identical ownership-marker name; the first ID is orphaned.

## Expected result

Before creating, the reconciler should recover an existing remote policy by its ownership-marker name (as `findAccessGroupByName` does for AccessGroup) or checkpoint the create intent durably before the remote call, so a lost `status.policyId` does not produce a duplicate remote policy.

## Relevant code

- Create path with no pre-create lookup: https://github.com/isac322/flareway/blob/ca23ba86f738214cc9ea7393656a74ef0782ace9/internal/controller/accesspolicy_controller.go#L145-L152
- `status.policyId` written only after the remote create: https://github.com/isac322/flareway/blob/ca23ba86f738214cc9ea7393656a74ef0782ace9/internal/controller/accesspolicy_controller.go#L442-L456
- Deletion keyed solely on `status.policyId`: https://github.com/isac322/flareway/blob/ca23ba86f738214cc9ea7393656a74ef0782ace9/internal/controller/accesspolicy_controller.go#L425-L441
- AccessGroup contrast — `findAccessGroupByName` recovery and `persistIdentity` retry: https://github.com/isac322/flareway/blob/ca23ba86f738214cc9ea7393656a74ef0782ace9/internal/controller/accessgroup_controller.go#L174, https://github.com/isac322/flareway/blob/ca23ba86f738214cc9ea7393656a74ef0782ace9/internal/controller/accessgroup_controller.go#L372-L380

## Acceptance criteria

- [ ] A lost `status.policyId` after `CreateAccessPolicy` does not produce a second remote policy; the existing remote policy is recovered by its ownership-marker name or an equivalent durable checkpoint.
- [ ] The attached probe fails on the fixed code path (it asserts the current defective behavior) and a fixed-path equivalent passes.
- [ ] `reconcileDelete` still removes exactly the remote policy the AccessPolicy owns.
